### ISSUE TITLE
Error in ONNX conversion 

### ISSUE THREAD

**Issue by ghoshsdeep** [rayguan97/GANav-offroad#18]
![Screenshot from 2022-09-15 16-15-11](https://user-images.githubusercontent.com/112154951/190384869-4c39b2b3-5501-4659-a87c-1173aa00e87f.png)
![Screenshot from 2022-09-15 16-14-22](https://user-images.githubusercontent.com/112154951/190384953-585e7f32-e2d2-430d-b81d-fa92fd2e3860.png)

I am trying to run the file pytorch2onnx.py using the pretrained model from https://drive.google.com/drive/folders/1Un-s7S3WjNTLjkhXPnOAK4RoUjL-pibk

Though the test.py runs ,pytorch2onnx gives a mismatch errror between the model and the loaded state dict.Can you please help me to resolve this issue?Am I using the proper config?

**Comment by rayguan97** [comment]
You would encounter this error when the image size from the data loader is not correct. To make sure the assertion go through correctly, you might need to:
1. modify the network parameters (self.h) and number of down-samplings, so that the size of the middle features would match up. (in which case you can't exactly load the pre-trained model)
2. Modify the input size (cropped size), in which case you can load the model. No matter the size of you input, as long as the cropped size is (300, 375) for rugd and (375, 600) for rellis, you should be fine.

Based on the information, can you figure out the issue more? What kind of data are you using, and what is the shape of the image from the data-loader? 

Can you try using the configs provided under trained_models, instead of the ones under configs?

**Comment by ghoshsdeep** [comment]
Thanks for your reply.I did use the config in trained models along with your downloaded pretrained model from https://drive.google.com/drive/folders/1Un-s7S3WjNTLjkhXPnOAK4RoUjL-pibk 
Also I changed the default size from (256,256) to (300,375) in the file **pytorch2onnx.py**.However while debugging I see that the same error pops up due to self.h = 10 and self.h =12 instead of them being 38 and 47 .
Upon inspection I see it happens from line 88 in **segmentors/encoder_decoder.py**,coming to line 160 in **decode_heads/ours_head_class_attn.py**,then to line 106 in the same
![Screenshot from 2022-09-20 13-31-53](https://user-images.githubusercontent.com/112154951/191227512-662cfbb3-b42b-405e-8c22-acd3b0505cdd.png)
 file.
As I see the image size is (300,375).
![Screenshot from 2022-09-20 13-33-18](https://user-images.githubusercontent.com/112154951/191227585-a6752c8c-c876-4a04-8940-7
![Screenshot from 2022-09-20 13-34-19](https://u
![Screenshot from 2022-09-20 13-40-27](https://user-images.githubusercontent.com/112154951/191227694-960e2d1e-bead-44bb-9f2b-7d0e0f835e03.png)
ser-images.githubusercontent.com/112154951/191227619-afb46e6e-8e20-439d-a6bb-c3311d3cb9b6.png)
54ce7cd240f.png)

**Comment by rayguan97** [comment]
Did you change the size or the cropped_size? You should make sure the cropped size is (300,375).

How it works is that in the config, model.decode_head.strides is an array/tuple, and the self.h and self.w is calculated based on that array(https://github.com/rayguan97/GANav-offroad/blob/main/mmseg/models/decode_heads/ours_head_class_attn.py#L82
). This same array also decided based on the structure of the backbone (number of downsampling and strides.) The code needs to be improve when I find some time.

For a quick solution, just hard code the fmap with desired dimension, but this would not load the pre-trained model well. If the cropped size are the same, it should run without problem. Did you have any issue running the original code and data?

**Comment by ghoshsdeep** [comment]
As far as I can see there are parameters for crop_size in **pytorch2onnx.py.**There is an **input size** which I think needs to be (300,375) in order to run so that self.h = 38 and self.w = 47 at all the instances.However it turns out to be **10 and 12 in one of the instances**
![Screenshot from 2022-09-22 15-14-55](https://user-images.githubusercontent.com/112154951/191718198-350a6b62-1f35-4cbc-b1c3-97cc437eb5bd.png)

The other files **train.py and test.py run** absolutely fine.
However the **pytorch2onnx.py.** does not even when i used your  pretrained model and config under trained models.
Was this file running succesfully for you?

**Comment by rayguan97** [comment]
I only used train.py and test.py for the code. I did not use pytorch2onnx.py, so I do not have support or experience for this. 

Are you saying that in most instance, it works fine but there is one case you have 10, 12? Are the inputs the same size? Could you check whether you added the padding? (which is specified in the config as well.)

**Comment by ghoshsdeep** [comment]
Yes thats correct, there is only one is instance.How did you convert it to ONNX format?How did you deploy this on the robot?
I see your name mentioned in the section of pytorch2onnx.py.You re T guan right ?Thats why I thought may be you would know .

**Comment by rayguan97** [comment]
That file is directly from the mmseg repo, which this code is built on.

To deploy, we put a laptop on the robot. But we are currently working on some mobile platform, and hopefully will update and share some instructions soon.

Can you try to remove that instance and see how it works? Or can you visualize that sample and see what’s the difference compared to other samples?
________________________________

**Comment by ghoshsdeep** [comment]
Actually I was planning to deploy it on Jetson Nano.Thats why I needed to export it as ONNX model.Are you also planning to do the same?

**Comment by rayguan97** [comment]
We are not currently working on that platform so I'm not very familiar with the conversion.

**Comment by ghoshsdeep** [comment]
Thank you for your kind support.I am closing this issue.If there is any other issues in future ,I would definitely like  share with you 
Once again Thank you for your help
